File: 0013-roman-to-integer.cpp
#include <algorithm>
#include <functional>
#include <string>
#include <map>
#include <vector>

class Solution {
public:
    int romanToInt(std::string s) {
        int result = 0;
        std::vector<int> values = {1000, 500, 100, 50, 10, 5, 1};
        std::vector<char> reprs = {'M', 'D', 'C', 'L', 'X', 'V', 'I'};

        for (int i = 0; i < s.size(); i++) {
            auto current_roman_index = std::find(reprs.begin(), reprs.end(), s[i]) - reprs.begin();
            if (i + 1 < s.size()) {
                auto next_roman_index = std::find(reprs.begin(), reprs.end(), s[i + 1]) - reprs.begin();
                if (values[current_roman_index] < values[next_roman_index]) {
                    result -= values[current_roman_index];
                }else {
                    result += values[current_roman_index];
                }
            }else {
                result += values[current_roman_index];
            }
        }
        return result;
    }
};

File: 0013-roman-to-integer.py
class Solution:
    def romanToInt(self, s: str) -> int:
        roman_identifier = {
            "M": 1000,
            "D": 500,
            "C": 100,
            "L": 50,
            "X": 10,
            "V": 5,
            "I": 1,
        }
        result = 0

        for index, ch in enumerate(s):
            if index + 1 < len(s) and roman_identifier[ch] < roman_identifier[s[index + 1]]:
                result -= roman_identifier[ch]
            else:
                result += roman_identifier[ch]
        return result

    def _romanToInt(self, s: str) -> int:
        roman_identifier = {
            "M": 1000,
            "CM": 900,
            "D": 500,
            "CD": 400,
            "C": 100,
            "XC": 90,
            "L": 50,
            "XL": 40,
            "X": 10,
            "IX": 9,
            "V": 5,
            "IV": 4,
            "I": 1,
        }
        result = 0
        for k, v in roman_identifier.items():
            while s.startswith(k):
                result += v
                s = s[len(k):]
        return result
